File: quickSort/quickSort.py
def quickSort(aList):

    _quickSort(aList, 0, len(aList) - 1)


def _quickSort(aList, first, last):

    if first < last:
        pivot = partition(aList, first, last)
        # print pivot

        _quickSort(aList, first, pivot - 1)
        _quickSort(aList, pivot + 1, last)


def partition(aList, first, last):
    pivot = aList[first]
    print 'pivot'
    print pivot
    left = first + 1
    right = last
    done = False

    while not done:
        while left <= right and aList[left] <= pivot:
            left += 1
        while left <= right and aList[right] >= pivot:
            right -= 1
        if right < left:
            # print ' ! '
            done = True
        else:
            swap(aList, left, right)
            # print aList

    print aList
    swap(aList, first, last)
    print aList
    print '!'
    return right


def swap(aList, x, y):
    temp = aList[y]
    aList[y] = aList[x]
    aList[x] = temp


test1 = [4, 6, 13, 6, 2, 7]
test2 = [5, 3, 135, 3, 6]
test3 = [1]

quickSort(test1)
# quickSort(test2)
# quickSort(test3)
# print quickSort(test1)
# print quickSort(test2)
# print quickSort(test3)

# print test1
# print test2
# print test3

File: quickSort/quickSort.js
function quickSort(arr) {
  _quickSort(arr, 0, arr.length - 1);
}

function _quickSort(arr, first, last) {

  if (first >= last) {

    return;

  } else {
    var pivot = partition(arr, first, last);

    _quickSort(arr, first, pivot - 1);
    _quickSort(arr, pivot + 1, last);
  }
}

function partition(arr, first, last) {
  var pivot = arr[first];
  var left = first + 1;
  var right = last;
  var done = false;

  while (!done) {
    while (left <= right && arr[left] <= pivot) {
      left++;
    }
    while (left <= right && arr[right] >= pivot) {
      right--;
    }
    if (right < left) {
      done = true;
    } else {
      swap(arr, left, right);
    }
  }

  swap(arr, first, last);
  return right;

}

function swap(arr, x, y) {
  var temp = arr[y];
  arr[y] = arr[x];
  arr[x] = temp;
}

test1 = [4, 6, 13, 6, 2, 7]
test2 = [5, 3, 135, 3, 6]
test3 = [1]

quickSort(test1)
quickSort(test2)
quickSort(test3)
  // console.log(quickSort(test1))
  // console.log(quickSort(test2))
  // console.log(quickSort(test3))

console.log(test1)
console.log(test2)
console.log(test3)

//quickSort(test2)